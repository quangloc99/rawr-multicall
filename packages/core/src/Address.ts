export type RawAddressString = string;
export type AddressString = {
    type: 'string';
    address: RawAddressString;
};

export type LabeledAddress = {
    type: 'labeled';
    label: string | symbol;
};

export type Address = AddressString | LabeledAddress;
export type AddressOrRawAddress = Address | RawAddressString;

export const labeledAddress = (label: LabeledAddress['label']): LabeledAddress => ({
    type: 'labeled',
    label,
});

export const castToAddress = (address: AddressOrRawAddress): Address =>
    typeof address == 'string' ? { type: 'string', address: address } : address;
