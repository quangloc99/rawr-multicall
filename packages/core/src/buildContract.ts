import { InstructionContext, PreprocessingInstructionContext } from './InstructionContext';
import { Bytes, concatBytes, bytesToHexWith0x } from './Bytes';
import { assert, assertDefined } from './errors';
import * as ins from './instructions';

export const DEFAULT_LABEL_SIZE = 3;

export type InstructionContextParams = {
    labelSize?: number;
    allowPUSH0?: boolean;
};

export function buildContract(
    instructions: ins.Instruction[],
    params?: InstructionContextParams
): {
    splittedByteCodes: Bytes[];
    byteCode: string;
} {
    const { labelSize = DEFAULT_LABEL_SIZE, allowPUSH0 = false } = params ?? {};
    const labelPosition = new Map<string, number>();
    let totalSize = 0;

    const preprocessingInstructionContext: PreprocessingInstructionContext = {
        getLabelSize: () => labelSize,
        allowPUSH0: () => allowPUSH0,
        addLabel(label: string) {
            assert(!labelPosition.has(label), `Duplicated label ${JSON.stringify(label)}`);
            labelPosition.set(label, totalSize);
        },
    };

    for (const instruction of instructions) {
        instruction.register?.(preprocessingInstructionContext);
        const insSize = instruction.byteSize(preprocessingInstructionContext);
        totalSize += insSize;
    }

    const instructionContext: InstructionContext = {
        ...preprocessingInstructionContext,

        instructions,
        getPos(label: string) {
            const res = labelPosition.get(label);
            return assertDefined(res, `Label ${JSON.stringify(label)} not found`);
        },
        getTotalSize: () => totalSize,
    };

    const splittedByteCodes: Bytes[] = [];
    for (const instruction of instructions) {
        splittedByteCodes.push(instruction.generate(instructionContext));
    }

    const byteCode = bytesToHexWith0x(concatBytes(splittedByteCodes));
    return {
        splittedByteCodes,
        byteCode,
    };
}
