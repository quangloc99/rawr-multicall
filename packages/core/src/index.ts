export * from './Address';
export * from './InstructionContext';
export * from './instructions';
export * from './Bytes';
export * from './Call';
export * from './buildContract';
export * from './buildRawMulticallContract';
export * from './decodeOutput';
export * from './errors';
export * from './registerPredeployContract';
export * from './Result';
