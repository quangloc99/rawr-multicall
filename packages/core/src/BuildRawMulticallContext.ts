import { Bytes } from './Bytes';
import { AddressOrRawAddress } from './Address';

export type BuildRawMulticallContext = {
    resolveAddress(address: AddressOrRawAddress): Bytes;
};
