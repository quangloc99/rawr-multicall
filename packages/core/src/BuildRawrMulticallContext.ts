import { Bytes } from './Bytes';
import { AddressOrRawAddress, LabeledAddress } from './Address';

export type BuildRawrMulticallContext = {
    getBuildingContractAddress(): Bytes;
    getLabeledAddressSalt(label: LabeledAddress['label']): Bytes;
    getLabeledAddress(label: LabeledAddress['label']): Bytes;
    resolveAddress(address: AddressOrRawAddress): Bytes;
};
