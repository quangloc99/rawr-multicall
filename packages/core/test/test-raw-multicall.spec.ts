import { Bytes, buildRawMulticallContract, createCall, decodeResult } from '../src';
import { Interface, JsonRpcProvider } from 'ethers';
import { describeForChain, CHAIN_ID_MAPPING, testData } from '@raw-multicall/test-helper';
import { ERC20__factory } from '@raw-multicall/test-helper/ethers-v6-contracts/typechain-types';
import { ERC20Interface } from '@raw-multicall/test-helper/ethers-v6-contracts/typechain-types/ERC20';

const ERC20Iface = new Interface(ERC20__factory.abi) as unknown as ERC20Interface;

describeForChain(
    (chain: number) => `buildRawMulticallContract for chain ${chain}`,
    (rpcUrl: string, chain: number) => {
        const CUR_TEST_DATA = testData.ERC20[chain];

        let provider: JsonRpcProvider;
        const blockTag = CUR_TEST_DATA.blockNumber;

        const doSend = (data: Bytes) => provider.call({ data, blockTag });
        const allowPUSH0 = chain != CHAIN_ID_MAPPING.ARBITRUM;

        beforeAll(() => {
            provider = new JsonRpcProvider(rpcUrl);
        });

        describe('erc20', () => {
            it('simple', async () => {
                const callData = ERC20Iface.encodeFunctionData('balanceOf', [CUR_TEST_DATA.holders[0]]);
                const calls = [createCall(CUR_TEST_DATA.tokenAddresses[1], callData)] as const;
                const sendData = buildRawMulticallContract(calls, { allowPUSH0 });
                expect(callData).toMatchSnapshot();
                expect(sendData).toMatchSnapshot();
                const res = await doSend(sendData.byteCode);
                expect(res).toMatchSnapshot();
                const output = decodeResult(calls, res);
                expect(output).toMatchSnapshot();
            });

            it('complex', async () => {
                const calls = CUR_TEST_DATA.tokenAddresses
                    .map((tokenAddress) => [
                        createCall(tokenAddress, ERC20Iface.encodeFunctionData('name')),
                        createCall(tokenAddress, ERC20Iface.encodeFunctionData('symbol')),
                        createCall(tokenAddress, ERC20Iface.encodeFunctionData('decimals')),
                        createCall(tokenAddress, ERC20Iface.encodeFunctionData('totalSupply')),
                        ...CUR_TEST_DATA.holders.map((holder) =>
                            createCall(tokenAddress, ERC20Iface.encodeFunctionData('balanceOf', [holder]))
                        ),
                    ])
                    .flat();

                const sendData = buildRawMulticallContract(calls, { allowPUSH0 });
                expect(sendData).toMatchSnapshot();
                const res = await doSend(sendData.byteCode);
                expect(res).toMatchSnapshot();
                const output = decodeResult(calls, res);
                expect(output).toMatchSnapshot();
            });
        });
    }
);
