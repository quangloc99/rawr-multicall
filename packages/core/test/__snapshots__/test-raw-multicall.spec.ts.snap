// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildRawMulticallContract for chain 1 erc20 complex 1`] = `
{
  "byteCode": "0x60586200024060803960ff60e0600460d85f5f600460805f73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb485af1841b3d841b01815281013d5f823e3d015f5f600460845f73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb485af1841b3d841b01815281013d5f823e3d015f5f600460885f73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb485af1841b3d841b01815281013d5f823e3d015f5f6004608c5f73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb485af1841b3d841b01815281013d5f823e3d015f5f602460905f73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb485af1841b3d841b01815281013d5f823e3d015f5f602460b45f73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb485af1841b3d841b01815281013d5f823e3d015f5f600460805f73808507121b80c02388fad14726482e061b8da8275af1841b3d841b01815281013d5f823e3d015f5f600460845f73808507121b80c02388fad14726482e061b8da8275af1841b3d841b01815281013d5f823e3d015f5f600460885f73808507121b80c02388fad14726482e061b8da8275af1841b3d841b01815281013d5f823e3d015f5f6004608c5f73808507121b80c02388fad14726482e061b8da8275af1841b3d841b01815281013d5f823e3d015f5f602460905f73808507121b80c02388fad14726482e061b8da8275af1841b3d841b01815281013d5f823e3d015f5f602460b45f73808507121b80c02388fad14726482e061b8da8275af1841b3d841b01815281013d5f823e3d0160d8900360d8f306fdde0395d89b41313ce56718160ddd70a08231000000000000000000000000a3a7b6f88361f48403514059f1f16c8e78d60eec70a0823100000000000000000000000028c6c06298d514db089934071355e5743bf21d60",
  "splittedByteCodes": [
    "0x6058",
    "0x62000240",
    "0x6080",
    "0x39",
    "0x60ff",
    "0x60e0",
    "0x6004",
    "0x60d8",
    "0x5f",
    "0x5f",
    "0x6004",
    "0x6080",
    "0x5f",
    "0x73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x5f",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x5f",
    "0x5f",
    "0x6004",
    "0x6084",
    "0x5f",
    "0x73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x5f",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x5f",
    "0x5f",
    "0x6004",
    "0x6088",
    "0x5f",
    "0x73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x5f",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x5f",
    "0x5f",
    "0x6004",
    "0x608c",
    "0x5f",
    "0x73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x5f",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x5f",
    "0x5f",
    "0x6024",
    "0x6090",
    "0x5f",
    "0x73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x5f",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x5f",
    "0x5f",
    "0x6024",
    "0x60b4",
    "0x5f",
    "0x73a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x5f",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x5f",
    "0x5f",
    "0x6004",
    "0x6080",
    "0x5f",
    "0x73808507121b80c02388fad14726482e061b8da827",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x5f",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x5f",
    "0x5f",
    "0x6004",
    "0x6084",
    "0x5f",
    "0x73808507121b80c02388fad14726482e061b8da827",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x5f",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x5f",
    "0x5f",
    "0x6004",
    "0x6088",
    "0x5f",
    "0x73808507121b80c02388fad14726482e061b8da827",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x5f",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x5f",
    "0x5f",
    "0x6004",
    "0x608c",
    "0x5f",
    "0x73808507121b80c02388fad14726482e061b8da827",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x5f",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x5f",
    "0x5f",
    "0x6024",
    "0x6090",
    "0x5f",
    "0x73808507121b80c02388fad14726482e061b8da827",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x5f",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x5f",
    "0x5f",
    "0x6024",
    "0x60b4",
    "0x5f",
    "0x73808507121b80c02388fad14726482e061b8da827",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x5f",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x60d8",
    "0x90",
    "0x03",
    "0x60d8",
    "0xf3",
    "0x",
    "0x",
    "0x06fdde0395d89b41313ce56718160ddd70a08231000000000000000000000000a3a7b6f88361f48403514059f1f16c8e78d60eec70a0823100000000000000000000000028c6c06298d514db089934071355e5743bf21d60",
  ],
  "totalValue": 0,
}
`;

exports[`buildRawMulticallContract for chain 1 erc20 complex 2`] = `"0x800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000855534420436f696e0000000000000000000000000000000000000000000000008000006000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000004555344430000000000000000000000000000000000000000000000000000000080000020000000000000000000000000000000000000000000000000000000000000000680000020000000000000000000000000000000000000000000000000005a9afec6a2aa60800000200000000000000000000000000000000000000000000000000000000000000000800000200000000000000000000000000000000000000000000000000000946e5d46833c800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000650656e646c650000000000000000000000000000000000000000000000000000800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000650454e444c45000000000000000000000000000000000000000000000000000080000020000000000000000000000000000000000000000000000000000000000000001280000020000000000000000000000000000000000000000000bfadbf953a8e088d5e9ce68000002000000000000000000000000000000000000000000011009602009976b7403ba5800000200000000000000000000000000000000000000000000d331ff1ff3cdae19b2127"`;

exports[`buildRawMulticallContract for chain 1 erc20 complex 3`] = `
[
  {
    "result": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000855534420436f696e000000000000000000000000000000000000000000000000",
    "success": true,
  },
  {
    "result": "0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000045553444300000000000000000000000000000000000000000000000000000000",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000000000000000000000006",
    "success": true,
  },
  {
    "result": "0x000000000000000000000000000000000000000000000000005a9afec6a2aa60",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000000000000946e5d46833c",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000650656e646c650000000000000000000000000000000000000000000000000000",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000650454e444c450000000000000000000000000000000000000000000000000000",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000000000000000000000012",
    "success": true,
  },
  {
    "result": "0x000000000000000000000000000000000000000000bfadbf953a8e088d5e9ce6",
    "success": true,
  },
  {
    "result": "0x00000000000000000000000000000000000000000011009602009976b7403ba5",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000d331ff1ff3cdae19b2127",
    "success": true,
  },
]
`;

exports[`buildRawMulticallContract for chain 1 erc20 simple 1`] = `"0x70a08231000000000000000000000000a3a7b6f88361f48403514059f1f16c8e78d60eec"`;

exports[`buildRawMulticallContract for chain 1 erc20 simple 2`] = `
{
  "byteCode": "0x60246200004660803960ff60e0600460a45f5f602460805f73808507121b80c02388fad14726482e061b8da8275af1841b3d841b01815281013d5f823e3d0160a4900360a4f370a08231000000000000000000000000a3a7b6f88361f48403514059f1f16c8e78d60eec",
  "splittedByteCodes": [
    "0x6024",
    "0x62000046",
    "0x6080",
    "0x39",
    "0x60ff",
    "0x60e0",
    "0x6004",
    "0x60a4",
    "0x5f",
    "0x5f",
    "0x6024",
    "0x6080",
    "0x5f",
    "0x73808507121b80c02388fad14726482e061b8da827",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x5f",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x60a4",
    "0x90",
    "0x03",
    "0x60a4",
    "0xf3",
    "0x",
    "0x",
    "0x70a08231000000000000000000000000a3a7b6f88361f48403514059f1f16c8e78d60eec",
  ],
  "totalValue": 0,
}
`;

exports[`buildRawMulticallContract for chain 1 erc20 simple 3`] = `"0x8000002000000000000000000000000000000000000000000011009602009976b7403ba5"`;

exports[`buildRawMulticallContract for chain 1 erc20 simple 4`] = `
[
  {
    "result": "0x00000000000000000000000000000000000000000011009602009976b7403ba5",
    "success": true,
  },
]
`;

exports[`buildRawMulticallContract for chain 42161 erc20 complex 1`] = `
{
  "byteCode": "0x60586200027060803960ff60e0600460d86000600060046080600073fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb95af1841b3d841b01815281013d6000823e3d016000600060046084600073fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb95af1841b3d841b01815281013d6000823e3d016000600060046088600073fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb95af1841b3d841b01815281013d6000823e3d01600060006004608c600073fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb95af1841b3d841b01815281013d6000823e3d016000600060246090600073fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb95af1841b3d841b01815281013d6000823e3d0160006000602460b4600073fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb95af1841b3d841b01815281013d6000823e3d0160006000600460806000730c880f6761f1af8d9aa9c466984b80dab9a8c9e85af1841b3d841b01815281013d6000823e3d0160006000600460846000730c880f6761f1af8d9aa9c466984b80dab9a8c9e85af1841b3d841b01815281013d6000823e3d0160006000600460886000730c880f6761f1af8d9aa9c466984b80dab9a8c9e85af1841b3d841b01815281013d6000823e3d01600060006004608c6000730c880f6761f1af8d9aa9c466984b80dab9a8c9e85af1841b3d841b01815281013d6000823e3d0160006000602460906000730c880f6761f1af8d9aa9c466984b80dab9a8c9e85af1841b3d841b01815281013d6000823e3d0160006000602460b46000730c880f6761f1af8d9aa9c466984b80dab9a8c9e85af1841b3d841b01815281013d6000823e3d0160d8900360d8f306fdde0395d89b41313ce56718160ddd70a08231000000000000000000000000b38e8c17e38363af6ebdcb3dae12e0243582891d70a08231000000000000000000000000f89d7b9c864f589bbf53a82105107622b35eaa40",
  "splittedByteCodes": [
    "0x6058",
    "0x62000270",
    "0x6080",
    "0x39",
    "0x60ff",
    "0x60e0",
    "0x6004",
    "0x60d8",
    "0x6000",
    "0x6000",
    "0x6004",
    "0x6080",
    "0x6000",
    "0x73fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x6000",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x6000",
    "0x6000",
    "0x6004",
    "0x6084",
    "0x6000",
    "0x73fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x6000",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x6000",
    "0x6000",
    "0x6004",
    "0x6088",
    "0x6000",
    "0x73fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x6000",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x6000",
    "0x6000",
    "0x6004",
    "0x608c",
    "0x6000",
    "0x73fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x6000",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x6000",
    "0x6000",
    "0x6024",
    "0x6090",
    "0x6000",
    "0x73fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x6000",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x6000",
    "0x6000",
    "0x6024",
    "0x60b4",
    "0x6000",
    "0x73fd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x6000",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x6000",
    "0x6000",
    "0x6004",
    "0x6080",
    "0x6000",
    "0x730c880f6761f1af8d9aa9c466984b80dab9a8c9e8",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x6000",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x6000",
    "0x6000",
    "0x6004",
    "0x6084",
    "0x6000",
    "0x730c880f6761f1af8d9aa9c466984b80dab9a8c9e8",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x6000",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x6000",
    "0x6000",
    "0x6004",
    "0x6088",
    "0x6000",
    "0x730c880f6761f1af8d9aa9c466984b80dab9a8c9e8",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x6000",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x6000",
    "0x6000",
    "0x6004",
    "0x608c",
    "0x6000",
    "0x730c880f6761f1af8d9aa9c466984b80dab9a8c9e8",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x6000",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x6000",
    "0x6000",
    "0x6024",
    "0x6090",
    "0x6000",
    "0x730c880f6761f1af8d9aa9c466984b80dab9a8c9e8",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x6000",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x6000",
    "0x6000",
    "0x6024",
    "0x60b4",
    "0x6000",
    "0x730c880f6761f1af8d9aa9c466984b80dab9a8c9e8",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x6000",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x60d8",
    "0x90",
    "0x03",
    "0x60d8",
    "0xf3",
    "0x",
    "0x",
    "0x06fdde0395d89b41313ce56718160ddd70a08231000000000000000000000000b38e8c17e38363af6ebdcb3dae12e0243582891d70a08231000000000000000000000000f89d7b9c864f589bbf53a82105107622b35eaa40",
  ],
  "totalValue": 0,
}
`;

exports[`buildRawMulticallContract for chain 42161 erc20 complex 2`] = `"0x800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000a546574686572205553440000000000000000000000000000000000000000000080000060000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000045553445400000000000000000000000000000000000000000000000000000000800000200000000000000000000000000000000000000000000000000000000000000006800000200000000000000000000000000000000000000000000000000001b91ca97c216a800000200000000000000000000000000000000000000000000000000000701c468c419e8000002000000000000000000000000000000000000000000000000000000efe3847eb84800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000650656e646c650000000000000000000000000000000000000000000000000000800000600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000650454e444c4500000000000000000000000000000000000000000000000000008000002000000000000000000000000000000000000000000000000000000000000000128000002000000000000000000000000000000000000000000010ac16ce5f49333b50845c8000002000000000000000000000000000000000000000000001eee0f293b6dc405a7874800000200000000000000000000000000000000000000000000000000000000000000000"`;

exports[`buildRawMulticallContract for chain 42161 erc20 complex 3`] = `
[
  {
    "result": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000a5465746865722055534400000000000000000000000000000000000000000000",
    "success": true,
  },
  {
    "result": "0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000045553445400000000000000000000000000000000000000000000000000000000",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000000000000000000000006",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000000000001b91ca97c216a",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000000000000701c468c419e",
    "success": true,
  },
  {
    "result": "0x00000000000000000000000000000000000000000000000000000efe3847eb84",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000650656e646c650000000000000000000000000000000000000000000000000000",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000650454e444c450000000000000000000000000000000000000000000000000000",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000000000000000000000012",
    "success": true,
  },
  {
    "result": "0x00000000000000000000000000000000000000000010ac16ce5f49333b50845c",
    "success": true,
  },
  {
    "result": "0x00000000000000000000000000000000000000000001eee0f293b6dc405a7874",
    "success": true,
  },
  {
    "result": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "success": true,
  },
]
`;

exports[`buildRawMulticallContract for chain 42161 erc20 simple 1`] = `"0x70a08231000000000000000000000000b38e8c17e38363af6ebdcb3dae12e0243582891d"`;

exports[`buildRawMulticallContract for chain 42161 erc20 simple 2`] = `
{
  "byteCode": "0x60246200004a60803960ff60e0600460a460006000602460806000730c880f6761f1af8d9aa9c466984b80dab9a8c9e85af1841b3d841b01815281013d6000823e3d0160a4900360a4f370a08231000000000000000000000000b38e8c17e38363af6ebdcb3dae12e0243582891d",
  "splittedByteCodes": [
    "0x6024",
    "0x6200004a",
    "0x6080",
    "0x39",
    "0x60ff",
    "0x60e0",
    "0x6004",
    "0x60a4",
    "0x6000",
    "0x6000",
    "0x6024",
    "0x6080",
    "0x6000",
    "0x730c880f6761f1af8d9aa9c466984b80dab9a8c9e8",
    "0x5a",
    "0xf1",
    "0x84",
    "0x1b",
    "0x3d",
    "0x84",
    "0x1b",
    "0x01",
    "0x81",
    "0x52",
    "0x81",
    "0x01",
    "0x3d",
    "0x6000",
    "0x82",
    "0x3e",
    "0x3d",
    "0x01",
    "0x60a4",
    "0x90",
    "0x03",
    "0x60a4",
    "0xf3",
    "0x",
    "0x",
    "0x70a08231000000000000000000000000b38e8c17e38363af6ebdcb3dae12e0243582891d",
  ],
  "totalValue": 0,
}
`;

exports[`buildRawMulticallContract for chain 42161 erc20 simple 3`] = `"0x8000002000000000000000000000000000000000000000000001eee0f293b6dc405a7874"`;

exports[`buildRawMulticallContract for chain 42161 erc20 simple 4`] = `
[
  {
    "result": "0x00000000000000000000000000000000000000000001eee0f293b6dc405a7874",
    "success": true,
  },
]
`;
